### PR TITLE
 Add GUI and CLI support for building own libgit2

### PR DESCRIPTION
**Pull request by tobiH94**
Closes se-passau/VaRA#316

### CODE REVIEW

File: varats/vara_manager.py [hunk 1]
```diff
@@ -293,8 +314,8 @@ def init_vara_build(path_to_llvm, build_type: BuildType,
             run_with_output(cmake, post_out)
 
 
-def build_vara(path_to_llvm: str, install_prefix: str, build_type: BuildType,
-               post_out=lambda x: None):
+def build_vara(own_libgit, path_to_llvm: str, install_prefix: str,
```
Comment (vulder) [inline varats/vara_manager.py]: could you add the type bool here

File: varats/vara_manager.py [hunk 2]
```diff
@@ -83,6 +83,9 @@ def setup_vara(init, update, build, llvm_folder, install_prefix, version,
             download_vara(llvm_folder, post_out=post_out)
             checkout_vara_version(llvm_folder, version,
                                   build_type == BuildType.DEV)
+            if not own_libgit:
```
Comment (vulder) [inline varats/vara_manager.py]: why do we init/update in the case we do not want our own libgit?

File: varats/vara_manager.py [hunk 3]
```diff
@@ -120,9 +123,11 @@ def setup_vara(init, update, build, llvm_folder, install_prefix, version,
             pull_current_branch(llvm_folder)
             pull_current_branch(llvm_folder + "tools/clang/")
             pull_current_branch(llvm_folder + "tools/VaRA/")
+            if not own_libgit:
```
Comment (vulder) [inline varats/vara_manager.py]: same

File: uicomponents/BuildMenu.ui [hunk 4]
```diff
@@ -206,6 +206,13 @@
         </layout>
        </widget>
       </item>
+      <item>
+       <widget class="QCheckBox" name="libgitButton">
+        <property name="text">
+         <string>Own Libgit</string>
```
Comment (vulder) [inline uicomponents/BuildMenu.ui]:
```suggestion
         <string>Build own libgit2</string>
```

File: varats/driver.py [hunk 5]
```diff
@@ -91,6 +93,8 @@ def build_setup():
                         nargs='?', default=llvm_src_dir)
     parser.add_argument("installprefix", default=llvm_install_dir, nargs='?',
                         help="Folder to install LLVM. (Optional)")
+    parser.add_argument("--ownlibgit", action="store_true", default=False,
```
Comment (vulder) [inline varats/driver.py]:
```suggestion
    parser.add_argument("--own-libgit2", action="store_true", default=False,
```

**Review by vulder — CHANGES_REQUESTED** [review]
This currently only allows us to active building our own libgit2, should we add support to deactivate it, meaning unset the cmake flag? I switch the flag on/off for testing but I'm not sure how common this case is.

Could you also please resolve the merge conflicts.
thanks

File: varats/vara_manager.py (hunk 2 above)
Comment (tobiH94) [reply varats/vara_manager.py]: Okay, I think I misunderstood. Which one is our own libgit? The one that is already installed on our own system, or the one that is provided by VaRA through downloading in external?
I thought it was the one installed on the system.

File: varats/driver.py [hunk 6]
```diff
@@ -108,7 +107,8 @@ def build_setup():
     vara_version = args.version if args.version is not None else CFG['version']
 
     setup_vara(args.init, args.update, args.build, args.llvmfolder,
-               args.installprefix, vara_version, build_type, update_term)
+               args.installprefix, args.ownlibgit, vara_version, build_type,
```
Comment (vulder) [inline varats/driver.py]:
```Traceback (most recent call last):
  File ".local/bin/vara-buildsetup", line 11, in <module>
    load_entry_point('VaRA-Tool-Suite', 'gui_scripts', 'vara-buildsetup')()
  File "git/vara-tool-suite/varats/driver.py", line 110, in build_setup
    args.installprefix, args.ownlibgit, vara_version, build_type,
AttributeError: 'Namespace' object has no attribute 'ownlibgit'
```

File: varats/gui/buildsetup_window.py [hunk 7]
```diff
@@ -90,7 +91,8 @@ def run(self):
         """
         Run, build an installs VaRA in a diffrent thread.
         """
-        vara_manager.build_vara(self.path_to_llvm,
+        vara_manager.build_vara(self.own_libgit,
```
Comment (vulder) [inline varats/gui/buildsetup_window.py]: We need a way to check if the submodule is setup. If the user initialzied vara without the submodule but ticks to build his own libgit2 and the builds vara, we only set the cmake flag, but do not initialize the submodule.

File: uicomponents/BuildMenu.ui [hunk 8]
```diff
@@ -206,6 +206,13 @@
         </layout>
        </widget>
       </item>
+      <item>
+       <widget class="QCheckBox" name="libgitButton">
+        <property name="text">
+         <string>Build own libgit2</string>
```
Comment (vulder) [inline uicomponents/BuildMenu.ui]: After changing this Label, we need to regenerate the ui_BuildMenu.py file, otherwise the change will not take effekt.

**Review by vulder — CHANGES_REQUESTED** [review]

File: varats/vara_manager.py [hunk 9]
```diff
@@ -146,6 +151,22 @@ def fetch_remote(remote, repo_folder=""):
             git["fetch", remote] & FG
 
 
+def init_all_submodules(folder):
+    """
+    Inits all submodules.
+    """
+    with local.cwd(folder):
+        git["submodule", "init"] & FG
```
Comment (vulder) [inline varats/vara_manager.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182)

File: varats/vara_manager.py [hunk 10]
```diff
@@ -146,6 +151,22 @@ def fetch_remote(remote, repo_folder=""):
...
+
+
+def update_all_submodules(folder):
+    """
+    Updates all submodules.
+    """
+    with local.cwd(folder):
+        git["submodule", "update"] & FG
```
Comment (vulder) [inline varats/vara_manager.py]: Codacy Issue found: [Statement seems to have no effect](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182)

File: varats/vara_manager.py (hunk 9 above)
Comment (vulder) [reply varats/vara_manager.py]:
```python
git("submodule", "init")
```
should do the trick

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/vara_manager.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182&bid=11339477)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 2
           

Complexity increasing per file
==============================
- varats/vara_manager.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182&bid=11339477)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- varats/vara_manager.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182&bid=11339477)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 3
           

Complexity increasing per file
==============================
- varats/vara_manager.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182&bid=11339477)

File: varats/vara_manager.py [hunk 11]
```diff
@@ -305,10 +328,25 @@ def build_vara(path_to_llvm: str, install_prefix: str, build_type: BuildType,
...
+                               post_out)
         b_ninja = ninja["install"]
         run_with_output(b_ninja, post_out)
 
+
+def verify_build_structure(own_libgit: bool, path_to_llvm: str,
+                           install_prefix: str, post_out=lambda x: None):
+    if (not get_cmake_var("VARA_BUILD_LIBGIT") or not os.path.exists(
+            path_to_llvm + "/tools/VaRA/external/libgit2/CMakeLists.txt")) and own_libgit:
+        init_all_submodules(path_to_llvm + "/tools/VaRA/")
```
Comment (vulder) [inline varats/vara_manager.py]: This initializes all sub modules even if we do not want our own libgit.

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- varats/vara_manager.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182&bid=11339477)

**Comment by vulder** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 2
- Added 1
           

Complexity increasing per file
==============================
- varats/vara_manager.py  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/app/VaRA/VaRA-Tool-Suite/pullRequest?prid=3099182&bid=11339477)

**Review by vulder — APPROVED** [review]

**Pull request merged**